Fix traction ascii data

tiannh7 opened the pull request:
Hi @gassmoeller,

Based on the issue https://github.com/geodynamics/aspect/issues/6248, I found that even when using the latest test branch, the problem with the spherical shell models is still not resolved. However, I noticed that the way traction is read in seems problematic: by default, only one component (pressure, i.e., the normal component) is loaded.

The issue is that if I remove the `normal_vector` parameter in `interface.cc`, although it does not cause errors inside `boundary_traction`, it leads to errors in the Stokes and melt modules. Therefore, I am unsure how best to keep, remove, or make the `normal_vector` parameter compatible. This also affects the input format because previously the default was:

```cpp
const double pressure = Utilities::AsciiDataBoundary<dim>::get_data_component(boundary_indicator,
                                                                                    position,
                                                                                    0);
      return -pressure * normal_vector;
...
Utilities::AsciiDataBoundary<dim>::initialize(boundary_ids, 1);
```
which I believe is inadequate. At least for spherical shells, using ASCII data to provide the full traction components (r, phi, theta) seems more convenient than describing traction via functions.

For now, I have kept the unused normal_vector parameter in the code, which might trigger a warning but avoids breaking other parts of the code.

I would appreciate your thoughts or suggestions on whether there is a better way to handle this compatibility issue or improve the traction input for spherical geometries.

Thank you very much for your help!

Best regards,
tiannh

gassmoeller commented on include/aspect/boundary_traction/ascii_data.h:
```diff
@@ -95,9 +95,16 @@ namespace aspect
         std::set<types::boundary_id> boundary_ids;
+
+        /**
+         * Whether to specify velocity in x, y, z components, or
```
```suggestion
         * Whether to specify traction in x, y, z components, or
```

gassmoeller commented on include/aspect/boundary_traction/ascii_data.h:
```diff
@@ -95,9 +95,16 @@ namespace aspect
 
 
-#endif
+#endif
```
please check our instructions for indentation [here](https://aspect-documentation.readthedocs.io/en/latest/user/extending/contributing.html#installing-astyle), your editor did not create a final newline character (which github dislikes).

tiannh7 replied:
> please check our instructions for indentation [here](https://aspect-documentation.readthedocs.io/en/latest/user/extending/contributing.html#installing-astyle), your editor did not create a final newline character (which github dislikes).

Ok, I'm very sorry, I realized that every time I use astyle 2.04 on my mac m4pro it always causes the last character or bracket to be repeated. I always format manually and cause some errors. I will try it on a linux desktop later.

gassmoeller commented on source/boundary_traction/ascii_data.cc:
```diff
@@ -140,4 +166,4 @@ namespace aspect
                                             "for dimension independent expressions.")
   }
-}
+}
```
Please update the documentation of this plugin in lines 139 to 167 to reflect your changes. Please also update the corresponding class documentation in ascii_data.h line 35.

gassmoeller reviewed (changes requested):
Thanks a lot for these improvements, I left a few comments on the code style, but I also have one functional request: Your changes would change existing functionality and therefore break some models (namely the ones just wanting to prescribe pressure at the boundary). Could you maybe introduce an input variable `Prescribe pressure instead of full traction` and set its default value to `true`. Then you could either initialize the `AsciiDataBoundary` class with 1 component (if the parameter is true), or `dim` components (if false). Then all existing models should continue to work, and you could still implement your new functionality just by switching this input parameter.

Finally, I think this is a significant enough change that you could add a changelog entry describing the improvements (see [here](https://aspect-documentation.readthedocs.io/en/latest/user/extending/contributing.html#changelog-entries) for how to do that.

Thanks again, I think this is a very useful addition.

gassmoeller commented:
I forgot one request in my earlier review: Please add a test that shows your new feature (full traction vector in spherical coordinates) works. The process of creating tests is described [here](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html), and you could use the test `chunk_lithostatic_pressure_2d.prm` as a starting point (or any other suitable test, your choice). This just ensures that we do not accidentally break this feature in the future.

tiannh7 commented:
Would it be appropriate to name the new test file `ascii_data_boundary_traction_3d_shell_spherical.prm`?  This seems more descriptive and fitting for the feature being tested.

If so, should I also add a simplified data sample file, e.g., `shell_3d_top.0.txt`, in the `aspect/data/boundary-traction/ascii-data/test` directory to accompany the test?

Please let me know if this sounds good or if you prefer a different naming or setup.

Thanks again for your guidance!

gassmoeller commented:
> Would it be appropriate to name the new test file ascii_data_boundary_traction_3d_shell_spherical.prm? This seems more descriptive and fitting for the feature being tested.

> If so, should I also add a simplified data sample file, e.g., shell_3d_top.0.txt, in the aspect/data/boundary-traction/ascii-data/test directory to accompany the test?

Yes to both, that would be the best way forward.

gassmoeller commented on include/aspect/boundary_traction/ascii_data.h:
```diff
@@ -95,6 +115,19 @@ namespace aspect
+        * Whether to prescribe pressure (default: true) or full traction vector (false)
+        * at the boundary. If true, only 1 component will be used for the boundary condition.
+        */
+        bool prescribe_pressure_instead_of_full_traction=true;
```
This is not wrong, but it is not our usual code style to put the initialization into the class header. Could you move the initialization into the initializer of the constructor in the source file? See e.g. how `boundary_traction/function.cc` does this. This is purely a change to maintain the visual consistency and layout of the code, there is nothing wrong with the code you wrote.

gassmoeller commented on source/boundary_traction/ascii_data.cc:
```diff
@@ -84,6 +102,22 @@ namespace aspect
+        {
+          prm.declare_entry ("Use spherical unit vectors", "false",
+                             Patterns::Bool (),
+                             "Specify velocity as r, phi, and theta components "
```
```suggestion
                             "Specify traction as r, phi, and theta components "
```

gassmoeller commented on source/boundary_traction/ascii_data.cc:
```diff
@@ -84,6 +102,22 @@ namespace aspect
+          prm.declare_entry ("Use spherical unit vectors", "false",
+                             Patterns::Bool (),
+                             "Specify velocity as r, phi, and theta components "
+                             "instead of x, y, and z. Positive velocities point up, east, "
```
```suggestion
                             "instead of x, y, and z. Positive tractions point up, east, "
```

gassmoeller reviewed (changes requested):
This PR is in almost ready, thank you :+1:. Just two minor stylistic changes left to address, and in order to make the tester pass you will need to update the output of the test that you added, see the end of [this page](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for how to do that, or let me know in case you need help.

While you are changing this PR anyway, could you update your `main` branch and then rebase this branch on top of it? We merged a PR in the meantime that should fix the tester `linux / tests (geodynamics/aspect-tester:focal-dealii-master...`, which is still failing on this branch. Let me know in case you need help with this, but it may be as simple as:
```
git checkout main
git pull upstream main        (assuming 'upstream' is the remote that points to geodynamics/aspect)
git checkout fix_traction_ascii_data
git rebase main
```

tiannh7 commented:
Thank you very much for your helpful feedback and suggestions!

I followed your instructions and rebased my branch `fix_traction_ascii_data` onto the latest `main` branch, but when I tried to push, I saw a lot of file changes. I wonder if I might have done something wrong during the rebase or push process.

Here are the relevant commands and messages I encountered:
```bash
$ git push origin fix_traction_ascii_data
To https://github.com/tiannh7/aspect.git
! [rejected] fix_traction_ascii_data -> fix_traction_ascii_data (non-fast-forward)
error: failed to push some refs to 'https://github.com/tiannh7/aspect.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

$ git rebase origin/fix_traction_ascii_data
warning: skipped previously applied commit 8362354f4
warning: skipped previously applied commit 7844bea04
hint: use --reapply-cherry-picks to include skipped commits
hint: Disable this message with "git config advice.skippedCherryPicks false"
Successfully rebased and updated refs/heads/fix_traction_ascii_data.
```

Could you kindly confirm whether I handled the rebase and push correctly? Or did I miss any important steps that caused the large number of file changes?

Thank you very much for your time and help!

gassmoeller commented:
Thanks for the quick update. You are right something went wrong during your rebase. I am not exactly sure what happened, it looks like the branch got mixed up with some changes in the main branch (maybe you have some changes in your main branch that are not in the official ASPECT main branch?). Anyway, I took the liberty to repair your branch and upload the changes to this PR (as maintainer I can change your PRs). If the testers succeed this should now be ready to merge.

Thanks again for the feature contribution!

gassmoeller reviewed (approved):
Ready to merge if the testers pass.

gassmoeller commented:
/rebuild

tiannh7 commented:
Thank you very much, @gassmoeller, for your kind help and for fixing the branch!  
I really appreciate you taking the time to review my PR and resolve the issues with the rebase.  
Your support made this process much smoother for me.

gassmoeller commented:
No worries, every first PR to a new project is full of little pitfalls, we are here to help out with that. Thanks again for the contribution!

Looks like the testers are happy, I am going to merge.

tiannh7 commented:
@gassmoeller Thank you very much for your guidance and for merging the PR!

I noticed that the comment about moving the initialization from the class header to the constructor initializer in the source file was not addressed during the merge. I recall making some related changes in my "Fix minor typos" commit.

I'm not sure whether I should fix this now or keep it in mind and address it in my next PR. What would you recommend?

Also, could you kindly share if the ASPECT development team prefers using **rebase** or **merge** workflows for incorporating contributions? I want to follow the best practices for future submissions.

Thanks again for your support and help!

gassmoeller commented:
> I noticed that the comment about moving the initialization from the class header to the constructor initializer in the source file was not addressed during the merge. I recall making some related changes in my "Fix minor typos" commit.

Ah sorry, I must have accidentally deleted that during my fixes.

> I'm not sure whether I should fix this now or keep it in mind and address it in my next PR. What would you recommend?

If you have a few minutes yes please open another PR. This one should be quick and easy as it is a single change.

> Also, could you kindly share if the ASPECT development team prefers using rebase or merge workflows for incorporating contributions? I want to follow the best practices for future submissions.

We usually try to do the following:
- develop all new features on feature branches with only a single new feature, try to start the feature branches from an up-to-date version of the main branch
- never merge from the main branch into feature branches
- feature branches can be merged into main even if they are not up-to-date with the latest version of main, our CI takes care to test the merged version as well
- if new features on main are needed in a feature branch (e.g. to resolve conflicts or use new features) then we ask to do a rebase of the feature branch to the latest main and then a force push to the Github PR
